Userland/Libraries/LibJS/Bytecode/Pass/LoadElimination.cpp

```diff
@@ -0,0 +1,193 @@
+                    // This only works, if the destination is not part of an array construction,
+                    // because these need a contigious range of registers,
+                    // in that case we re-emit the store
+                    // FIXME: We might be able to elide the prevous load in the non-array case,
+                    //        because we do not yet reuse the accumulator
+                    auto const& store = static_cast<Op::Store const&>(next_instruction);
+
+                    if (array_ranges.get(store.dst().index())) {
+                        // The register is part of an array, so the store to it cannot
+                        // be simply replaced by a reference to another register
+                        new (new_block->next_slot()) Op::Store(store);
+                        new_block->grow(sizeof(Op::Store));
+                    } else {
+                        register_rerouteing_table.set(store.dst().index(), reg->value);
+                    }
+
+                    ++it;
+                }
+
+                continue;
+            }
+            new (new_block->next_slot()) Op::GetVariable(get_variable);
+            new_block->grow(sizeof(Op::GetVariable));
+
+            if (next_instruction.type() == Instruction::Type::Store) {
+                // If we haven't seen the currently loading variable and the next instruction is a store,
+                // We can cache the register we loaded the variable into, reusing that
+                // on later loads, to avoid looking going through the whole lookup chain
+                auto const& store = static_cast<Op::Store const&>(next_instruction);
+                identifier_table.set(get_variable.identifier().value(), store.dst());
+
+                new (new_block->next_slot()) Op::Store(store);
+                new_block->grow(sizeof(Op::Store));
+                ++it;
+            }
+
+            continue;
+        }
+        case SetVariable: {
+            // When a variable is deleted we need to remove it from the cache, because
+            // That is not valid anymore
+            auto const& set_variable = static_cast<Op::SetVariable const&>(*it);
+
+            identifier_table.remove(set_variable.identifier().value());
+
+            break;
+        }
+        case DeleteVariable: {
+            // When a variable is deleted we need to remove it from the cache, it does not
+            // exist anymore, although a variable of the same name may exist in upper scopes
+            auto const& set_variable = static_cast<Op::DeleteVariable const&>(*it);
+
+            identifier_table.remove(set_variable.identifier().value());
+
+            break;
+        }
+        case Store: {
+            // If we store to a position that we have are rerouteing from,
+            // we need to remove it from the routeing table
+            // FIXME: This may be redundant due to being SSA like in our register usage
```

Comment:
The bytecode is not in SSA form.

Reply:
SSA-like,
I meant, that we do not reuse registers,
will adjust the comment to be more explicit